actions/microsoft-calendar/actions.py

```diff
@@ -0,0 +1,213 @@
+from typing import Literal
+
+import requests
+from sema4ai.actions import OAuth2Secret, Response, action
+
+from models import Calendar, CreateEvent, Event, QueryParams, UpdateEvent
+
+
+def _build_headers(token, timezone=None):
+    headers = {
+        "Authorization": f"Bearer {token.access_token}",
+        "Content-Type": "application/json",
+    }
+
+    if timezone:
+        headers["Prefer"] = f'outlook.timezone="{timezone}"'
+
+    return headers
+
+
+@action(is_consequential=True)
+def create_event(
+    credentials: OAuth2Secret[
+        Literal["microsoft"],
+        list[Literal["Calendars.ReadWrite"]],
+    ],
+    event: CreateEvent,
+    calendar_id: str = "",
+) -> Response[Event]:
+    """Creates a new event in the user's calendar.
+
+    Args:
+        credentials: JSON containing Microsoft OAuth2 credentials.
+        event: JSON representation of the Microsoft Calendar event.
+        calendar_id: Calendar identifier where to add the event which can be found by listing calendars action.
+            Default value is using the user's default calendar.
+
+    Returns:
+        The newly created event.
+    """
+    headers = _build_headers(credentials)
+
+    url = "https://graph.microsoft.com/v1.0/me/calendar/events"
+    if calendar_id:
+        url = f"https://graph.microsoft.com/v1.0/me/calendars/{calendar_id}/events"
+
+    response = requests.post(
+        url,
+        headers=_build_headers(credentials),
+        json=event.model_dump(mode="json", exclude_none=True, exclude={"timeZone"}),
+    )
+
+    if response.status_code != 201:
+        return Response(error=response.text)
+
+    return Response(result=Event.model_validate(response.json()))
+
+
+@action(is_consequential=True)
+def update_event(
+    credentials: OAuth2Secret[
+        Literal["microsoft"],
+        list[Literal["Calendars.ReadWrite"]],
+    ],
+    event_id: str,
+    updates: UpdateEvent,
+) -> Response[Event]:
+    """Update an existing Microsoft Calendar event.
+
+    Args:
+        credentials: JSON containing Microsoft OAuth2 credentials.
+        event_id: Identifier of the event to update. Can be found by listing events in all calendars.
+        updates: A dictionary containing the event attributes to update.
+
+    Returns:
+        Updated event details.
+    """
+
+    headers = _build_headers(credentials)
+    updates_json = updates.model_dump(
+        mode="json", exclude_none=True, exclude={"timeZone"}
+    )
+
+    # We add the new attendees to the existing ones, so we don't override them
+    if "attendees" in updates_json:
+        response = requests.get(
+            f"https://graph.microsoft.com/v1.0/me/events/{event_id}", headers=headers
+        )
+        if not response.status_code == 200:
+            return Response(error=response.text)
+
+        updates_json["attendees"] = updates_json["attendees"] + response.json().get(
+            "attendees", []
+        )
```

Comment:
It might be an idea to make sure that, besides having a good response code, we get a parsable JSON before attempting to access the operational fields